Cleanup compaction_stop_exception

bhalevy opened the pull request:
The gist of this series is splitting `compaction_abort_exception` from `compaction_stop_exception`
and their respective error messages to differentiate between compaction being stopped due to e.g. shutdown
or api event vs. compaction aborting due to scrub validation error.

While at it, cleanup the existing retry logic related to `compaction_stop_exception`.

Test: unit(dev)
Dtest: nodetool_additional_test.py:TestNodetool.{{scrub,validate}_sstable_with_invalid_fragment_test,{scrub,validate}_ks_sstable_with_invalid_fragment_test,{scrub,validate}_with_one_node_expect_data_loss_test} (dev, w/ https://github.com/scylladb/scylla-dtest/pull/2267)

bhalevy commented:
Cc @avikivity, this addresses https://github.com/scylladb/scylla/pull/9315#discussion_r705469154

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -1275,7 +1275,7 @@ class scrub_compaction final : public regular_compaction {
     private:
```
patch title should describe what it's being done instead, like: drop retry in compaction_stop_exception

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -548,7 +545,7 @@ inline bool compaction_manager::maybe_stop_on_error(future<> f, stop_iteration w
         do_stop();
     } catch (...) {
-        cmlog.error("compaction failed: {}: {}", std::current_exception(), decision_msg);
+        cmlog.error("compaction failed: {}: retrying", std::current_exception());
```
the error msg is correct because will_stop overrides retry

bhalevy replied:
It is misleading to set `retry` to true in the following statement and return it later on.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -600,7 +600,7 @@ void compaction_manager::submit(column_family* cf) {
                 task->compaction_running = false;
 
                 if (!can_proceed(task)) {
-                    maybe_stop_on_error(std::move(f), stop_iteration::yes);
```
the reason why we had stop_iterator is because the caller can tell in advance if it wants a retry or not, like when compaction manager is being stopped.

raphaelsc commented on sstables/exceptions.hh:
```diff
@@ -45,14 +45,29 @@ struct bufsize_mismatch_exception : malformed_sstable_exception {
+class compaction_abort_exception : public compaction_job_exception {
+public:
+    compaction_abort_exception(sstring ks, sstring cf, sstring reason)
+        : compaction_job_exception(format("Compaction for {}/{} aborted: {}", ks, cf, reason)) {}
```
shouldn't the message follow a pattern as above, i.e. was aborted due to: ... ?

bhalevy replied:
yes, it can

raphaelsc commented on sstables/exceptions.hh:
```diff
@@ -45,14 +45,29 @@ struct bufsize_mismatch_exception : malformed_sstable_exception {
 
+// Indicates that compaction was stopped via an external event,
+// E.g. shutdown or api call.
+class compaction_stopped_exception : public compaction_job_exception {
```
if we're switching to past tense, perhaps abort should be aborted instead, for symmetry.

bhalevy replied:
ok

bhalevy commented:
In v2:
- Added "compaction_manager: move errors stats accounting to maybe_stop_on_error"
  - fixes #9354
- "compaction: drop retry parameter from compaction_stop_exception"
  - fixed patch title
- "compaction_manager: maybe_stop_on_error: sync return value with error message."
  - just make the return value match the `will_stop` param.
  - and simplify the printout.
- Dropped "compaction_manager: maybe_stop_on_error: drop will_stop parameter"
- renamed `compaction_abort_exception` -> `compaction_aborted_exception`
  - and made its `what()` message more similar to `compaction_stopped_exception`

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -550,10 +550,12 @@ inline bool compaction_manager::maybe_stop_on_error(future<> f, stop_iteration w
         decision_msg = !retry ? stop_msg : decision_msg;
         cmlog.info("compaction info: {}: {}", e.what(), decision_msg);
     } catch (storage_io_error& e) {
+        _stats.errors++;
```
thanks for fixing this

raphaelsc reviewed (approved):
LGTM

bhalevy commented:
@scylladb/scylla-maint please merge

bhalevy commented:
@scylladb/scylla-maint ping

denesb commented:
Queued as 61e7d3de904927340113ff5118f906b7da8ad76f.
